### PR TITLE
Add djgpp as CI cross-compile target

### PR DESCRIPTION
**Pull request by jwt27**
As the title says, this patch introduces CI builds for djgpp.  This will help
catch incompatibilities from future commits early, making this port easier to
maintain.

The djgpp toolchain is not available from any official Debian/Ubuntu
repositories, so instead it is downloaded and installed from a Launchpad PPA
that I personally maintain.  This is located here:

https://launchpad.net/~jwt27/+archive/ubuntu/djgpp-toolchain

Of course, this could introduce possible security risks.  If write permissions
are ever given in `cross-compiles.yml`, or any secrets are ever exported to the
build environment, software from this repository could gain access to that, so
care must be taken to avoid this.

I am leaving this PR as a draft for now, since the library currently does not
compile with `--strict-warnings`.  In order for this patch to be useful, that
would have to be resolved first.  The output from `make -k 2> error.log` may be
found here:

https://gist.github.com/jwt27/c7ac4f9b3f8c277c170e8acf374e8a5e

### CODE REVIEW

**Comment by t8m** [comment]
**OTC: do we want this CI build given the conditions mentioned above?**

**Comment by mattcaswell** [comment]
> OTC: do we want this CI build given the conditions mentioned above?

OTC: This PR would be acceptable if djgpp was a community maintained platform

@jwt27 - are you willing to be the maintainer for this platform?

https://www.openssl.org/policies/platformpolicy.html

**Comment by jwt27** [comment]
Sure, I suppose I can do that.

I also want to ping @jmgdjgpp and @gvanem - I see they have submitted patches
here before, and I know they are much more knowledgeable than me regarding
djgpp/watt32.  Maybe they would also be interested in helping out.

**Comment by jmgdjgpp** [comment]
Am Tue, 04 Oct 2022 10:32:37 -0700
schrieb jwt27 ***@***.***>:

> Sure, I suppose I can do that.
>
> I also want to ping @jmgdjgpp and @gvanem - I see they have submitted
> patches here before, and I know they are much more knowledgeable than
> me regarding djgpp/watt32.  Maybe they would also be interested in
> helping out.
>
>

I will support you as best as I can concerning DJGPP specific issues.
For Watt-32 gvanem is the better address because he is the maintainer
of the library.

Regards,
Juan M. Guerrero

**Comment by jwt27** [comment]
Force-push to re-trigger the build.  With #19322 merged, everything should work
now.  I also added a configure flag to disable atomics, which is necessary for
`-march=i386`.  Maybe that should go in the default configuration for djgpp.

@jmgdjgpp - Thanks for your response!  If I ever run into any difficult issues,
I'll let you know.

And sorry for the delay - #19322 took a while, and I also got sidetracked with
other stuff for a bit.

**Comment by t8m** [comment]
> I also added a configure flag to disable atomics, which is necessary for
> `-march=i386`. Maybe that should go in the default configuration for djgpp.

Is this necessary even with no-threads? What are the build failures?

And yes, if there is no way this would work with djgpp, it should be added to the djgpp default target instead of adding it to the default build.

**Comment by jwt27** [comment]
> > I also added a configure flag to disable atomics, which is necessary for
> > `-march=i386`. Maybe that should go in the default configuration for djgpp.
> 
> Is this necessary even with no-threads? What are the build failures?
> 
> And yes, if there is no way this would work with djgpp, it should be added to the djgpp default target instead of adding it to the CI build steps.

It's not necessarily a djgpp-specific issue.  Atomic instructions (`cmpxchg`) are not available on 386, so gcc compiles those to libatomic calls.  That library is not available on all targets (including djgpp), so you'll get undefined references on linking.

Using `-march=i486` is another solution, but personally I would prefer to maintain 386 compatibility.

I see atomics are only used for refcounts, so I suppose it's safe to disable them entirely when `no-threads` is given.

**Comment by jwt27** [comment]
Force-pushed again to re-trigger build - all should work now.

**Comment by t8m** [comment]
> Force-pushed again to re-trigger build - all should work now.

No need to do that. The buildbot CI sometimes fails on some workers due to intermittent problems with the workers.

**Comment by t8m** [comment]
> No need to do that. The buildbot CI sometimes fails on some workers due to intermittent problems with the workers.

Oh, I misunderstood, the force push was to retrigger the build after merging https://github.com/openssl/openssl/pull/19751. Ok then.

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
Needs a rebase now.

**Comment by jwt27** [comment]
Done, thanks for the reminder.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by tmshort** [comment]
Merged to master and openssl-3.1. Thank you for your contribution!

**Comment by jwt27** [comment]
Thanks!  I hope this will help keep the djgpp port alive and up to date.

**Pull request closed without merging**
